crates/ruff_linter/src/noqa.rs

```diff
@@ -178,14 +187,20 @@ impl Ranged for All {
 #[derive(Debug)]
 pub(crate) struct Codes<'a> {
     range: TextRange,
+    #[allow(clippy::struct_field_names)]
     codes: Vec<&'a str>,
+    code_ranges: Vec<TextRange>,
```

Comment:
I think it's fine to use two vectors here but we could also use a single vector if we want. However, I think it's important that we don't leak the fact that `codes` and `code_ranges` must contain exactly the same number of elements (which it currently does because the rule implementation uses `zip`).

I would change the implementation to 

```rust
impl<'a> Codes<'a> {
    /// The codes that are ignored by the `noqa` directive.
    pub(crate) fn names(&self) -> impl Iterator<Item=&'a str> + '_ {
        self.codes.iter().map(|(code, _)| *code)
    }

    pub(crate) fn iter(&self) -> std::slice::Iter<(&str, TextRange)> {
        self.codes.iter()
    }

    /// Returns `true` if the string list of `codes` includes `code` (or an alias
    /// thereof).
    pub(crate) fn includes(&self, needle: Rule) -> bool {
        let needle = needle.noqa_code();

        self.names()
            .any(|code| needle == get_redirect_target(code).unwrap_or(code))
    }

    pub(crate) fn ranges(&self) -> impl Iterator<Item=TextRange> + '_ {
        self.codes.iter().map(|&(_, range)| range)
    }
}
```

and then only use these methods in the calling code. It's then up to the implementation to decide if it wants to use a `Vec` storing a tuple for each code or two vectors (and I leave that decision to you). 



(Note I moved the `includes` function to `Codes`. That means we should delete the `noqa::includes` function and instead use `codes.includes(rule)`)

Reply:
Makes sense. Done.